### DISCUSSION TITLE
Safe exposing of reset-password site

### DISCUSSION THREAD

**Discussion by Sblop** [lldap/lldap#876] [Q&A] [answered]
Hi All,

I do not feel confident in exposing the main lldap site, as this is the backbone in my auth flow, using authelia with oauth etc. I do however need my users being able to access the reset password feature. The lldap is currently on availabe for the admins.

I initially tried blocking all of lldap-ish.domain.tdl except lldap-ish.domain.tdl/reset-password, howvever several recourses are loaded from the root /.

After going through the loaded resources, I have exposed below, without any auth. 

(Snip from Authelia config)
```
      resources:
      - "^/reset-password([/?].*)?$"
      - "^/static([/?].*)?$"
      - "^/pkg/lldap_app([/?].*)?$"
      - "^/auth/reset/step1([/?].*)?$"
```

I cannot however figure out if this is safe. I do not like exposing ANY of lldap, however as I need to, I like a second opinion on what to expose. I wish to obviously expose as little as possible, however the reset password functionality is quire essential.

Thanks in advance.

/Benjamin

**Comment by nitnelave** [comment, marked as answer]
Hey! It _should_ be fairly safe to expose LLDAP to the internet. Now, I understand that this is not much reassurance if you're cautious.

I see a few ways to only expose the necessary functionality:
- block all but the necessary resources, as you did. It works now, and your list looks good, but that's not exactly an intended use case so I can't guarantee that it'll keep working. You might end up with half broken features because we added a new query.
- block all post requests except to the reset password flow. In particular, if you block the login, it'll be hard for anyone to find a security flaw. If you allow the login, then users can see their own account.
- block the entire website and provide password resets through LDAP with a third-party website. In particular, I believe authelia can do password resets? Not 100% sure about that.

**Reply by jakob42** [reply]
Authelia can do password reset (I'm using it against a samba PDC). (But is it really safer than lldap?)

@Sblop I don't know how big your setup is. But if it doesn't need to be fully automated you could just put basic auth in front of your lldap backend from externally and hand out the password for the seldom password reset.

Personally I feel like lldap is safe enough to be accessible from the internet. There are services I host for myself and friends I have much less faith in.

**Reply by nitnelave** [reply]
Putting basic auth in front of a password reset form would be kinda missing the point :D

**Reply by Sblop** [reply]
Hi Nitnelave, Thanks for your incredible quick answer. Okay, I did not do my homework good enough... Yes, it turns out that Authelia is able to change the lldap password, thus no need to expose lldap to the internet. 
Reason I do not like to expose lldap is lack of 2 factor auth on the admin users. - I like keepling lldap basically only backend.

Again thanks for your quick replies. :-)

Just for reference, the authelia configuration is dead simple:

```
authentication_backend:
  refresh_interval: '5m'
  password_reset:
    disable: false
```

As long as the ldap connection is set, it just works, as long as the password reset has not been disabled.

**Reply by jakob42** [reply]
@nitnelave: Sure, in a big setup. But if its only for family and friends one could share that password on demand.

**Reply by nitnelave** [reply]
Ah, right a fixed "known" password, not an LDAP-based basic auth with your own password :D That makes sense!

**Comment by viggy96** [comment]
@nitnelave Can you provide a definitive guide on what parts of LLDAP to minimally expose for password reset? And maybe at least provide warnings on release if changes in that list are made on future updates to LLDAP?

**Reply by nitnelave** [reply]
So, just to make sure I understand: you want to directly expose LLDAP to your users so they can reset their passwords through the web interface, but nothing else?

That seems like a very narrow set of requirements, there are many other ways of doing password resets if you don't want to expose LLDAP. You can go through LDAP(S), use another service in front (Authelia, Jellyfin, ...)
Or just let your users access the frontend.

If you're set on exposing LLDAP but just for password resets, a very good approximation is to disallow access to the graphql API (though I haven't tried it, I'm not 100% sure it's compatible with the password reset flow).

That lets users reset their passwords, and technically also login, but they won't be able to pull (or push) any info to LLDAP.

**Reply by viggy96** [reply]
Is setting the "Password reset URL" in Jellyfin necessary to make that work? Its not in the example. And I'm using Pocket ID with LLDAP, which doesn't have LDAP password resetting functionality.

**Reply by nitnelave** [reply]
If Jellyfin is connected to LLDAP via the ldapauth plugin, it should be able to reset the password through LDAP, so you don't have to specify a password reset URL.

**Reply by viggy96** [reply]
And yeah I'm only exposing the frontend, not the underlying LDAP port. I'm getting some strange issue with some request being blocked preventing the password reset button working on Jellyfin, but it works fine with LLDAP directly of course.

**Reply by nitnelave** [reply]
Feel free to create an issue and include LLDAP verbose logs, so we can see what's going on.
